### ISSUE TITLE
Constants replaced by globalvariable addresses

### ISSUE THREAD

**Issue by daatno** [avast-tl/retdec#485]
LLVM converts llvm::ConstantInt values to GlobalVariable addresses, and it appears to me, that I can do nothing about it. Consider the following, shortened code:
...
0x18:    r2 = memread(0x4000)
0x1C:    r3 = memread(0x4004)
0x20:    r4 = add(r2, r3)
...
Gets translated to:
int32_t v1 = *(int32_t *)0x4004 + *(int32_t *)0x4000;
which is absolutely fine, however, in another binary:
...
0x800004:    r2 = memread(0x4000)
0x800008:    r3 = memread(0x4004)
0x80000C:    r6 = 0x4020
0x800010:    r1 = logic_shift_right(r6, 3)    
0x800014:    r4 = add(r2, r3)
...
Gets translated to:
int32_t g1 = 0;
int32_t g2 = 0;
int32_t g3 = 0;

// Address range: 0x800000 - 0x800040 
int32_t entry_point(void) {
    uint32_t v1 = g1 + g2;
    int32_t v2 = (int32_t)&g3 >> 3;
...
The funny thing is that now the address range is nowhere near 0x4000(just like in the previous example). But LLVM uses global variables, which happen to have the correct addresses, which can be looked up in the IR file. Another funny thing is, that for a bit shifting operation a global variable address is used as operand.

My problem is the following: 
I can trick LLVM to interpret memread(#address) instruction always as an instruction handling a constant, by converting the constant to a negative value and storing this info till the IR will be converted to high level language. However I cannot get LLVM to leave an llvm::Value as ConstantInt, if I pass it to IRBuilder for processing(for example through a register storing instruction).
Do you know how I can tell IRBuilder, or any other part of LLVM, to leave a supplied or retrieved llvm::Value as const?

**Comment by s3rvac** [comment]
@PeterMatula Can you please look into this?

**Comment by daatno** [comment]
Thanks. 
In the meantime I've found the constant->global converting code part in bin2llvmir, so I could set a custom conversion there. Previously I saw a case where a constant didn't show up in the checkForGlobalInInstruction(...) function(and had no means correcting it in bin2llvmir), but I did modify  IRBuilder usage since, and can't remember which case triggered such behaviour. 
So now the question is, how can I add info to an llvm::LoadInst for example(which references given constant) so that I can retrieve given info from the llvm::ConstantInt in bin2llvmir, before it gets converted?
